### DISCUSSION TITLE
How to use Gizmo

### DISCUSSION THREAD

**Discussion by tingspain** [zalo/CascadeStudio#69] [Q&A]
Hello, 

How to enable the use of Gizmo on the view? I tried the X, W, E, and R keys but it doesn't work. 

Could you provide an example of how to use it?

Thanks in adavance

**Comment by Irev-Dev** [comment]
@tingspain I think Gizmo is just for "click and drag" functionality for moving the camera with the mouse. scrolling also zooms in. I'm no away of any other hotkeys.
I might be wrong though.

**Reply by tingspain** [reply]
Of course, Gizmos (well gizmos is more than that in general terms), but Transformation Gizmos are for modifying the selected element using the mouse. There is one GIF showing the use of them at the main page, but I dont know how to activate them, that is why I tried different shortcuts to activate the Gizmo, but I didnt have any luck. 

![image](https://user-images.githubusercontent.com/1858083/107104787-16271d80-681b-11eb-9efe-1195aa6857a9.png)

**Comment by tingspain** [comment]
Ok, looking around the code, I can see that the way to use the Gizmos is by applying the function Transform to the shape. However, it doesn't work well for me. I don't know if I am doing something wrong, or it is just some bugs because it is in beta. 
Also, I noticed that the short keys are off for me, instead of being R for Rotate, T for Translate, S for Scale, the short keys are mapped to W for Translate, E for Rotate, and R for Scale. I am using an Intel MacBook Pro and macOS 11.2.1, 

This is how I use it:

```
let shape = Translate([-25, 0, 40], Text3D("Hi!"));
let trans = Transform(shape);
```

**Comment by raydeleu** [comment]
I have seen the Gizmo in my earlier attempts, but lately the gizmo no longer appears when I use the function Transform. In the developer window I noticed the following error message: 

```
TransformControls.js:15 Uncaught TypeError: Class constructor Pe cannot be invoked without 'new'
    at new THREE.TransformControls (TransformControls.js:15)
    at CascadeEnvironment.<anonymous> (CascadeViewHandles.js:10)
    at Worker.cascadeStudioWorker.onmessage (?THREE.TransformControls @ TransformControls.js:15
(anonymous) @ CascadeViewHandles.js:10
cascadeStudioWorker.onmessage @ ?
```

As I do not see the point of the gizmo in a code-based CAD system I do not miss this functionality, but I was unable to produce screenshots for the manual. Perhaps omit it completely from the manual as it does not work? Or am I doing something wrong?

**Reply by zalo** [reply]
You probably aren't missing anything; the Gizmos were more of a hack to see if I could get in-window interactions to bidirectionally edit the code.

Unfortunately, it relies on generating error messages in the backend, and parsing the stack-traces of those error messages in a super brittle way between browsers.   It's probably worth leaving out of any documentation.

**Reply by zalo** [reply]
Also, the transform gizmo implementation was broken in a recent update; [here's a working version now.](https://zalo.github.io/CascadeStudio/?code=fZLBahsxEIbvforfOWlhnbXjuKQNpRi3TQs%2BhNiQwNKDVp71ishSkOQ0Tum7d7QbL0kL1UHSaGb%2BT5pRUWC%2BvP42%2F4C1lzbUzu9wpZ93DmsKEStFlgZFgXVD%2F0TogBC1MdAW0mJuHhqJVZSRhinle0QjA6yLqIgsIgvSpgs%2BoPLuZyCPah%2BxaLzbEcTFLGPfBl%2B1p9o9wVncarvhwCFEciipGm23nEzyPkDHYTZ4QSm%2BQXWAOijDkMiK%2B22DR%2Bm12wfs3IYCHrXELUT7DiOjdjbL8QXixsWjdQOxUtIQbxPwDmLp2C6ujKukyU5b3GeqtSVM%2BXaeZaPD0tXxCB0Yiqilogk%2B4r%2Fj2pnD1llRlqOzWY52Gv%2FIUeKtmTYvRhfZGeyKfk%2FZZU88Y%2BLxdSTKcZ6iz5NG%2B0Q%2BmnBip3rOGq8vwMejSU98Y43%2FJmavmNMjM%2F2MFjCavcsxm6TMcjQ5HSdgP%2FPp%2B1m3dq6F9twykYid9mVb47aiEsEZvUFo5AP1TY38F5V3gb8fqdS3MNA1xKIhdV%2B5J3GScj%2BdHPV%2BtVqibFuSd3XqlmlfQfwe%2FAE%3D&gui=q1ZKzs8tyM9LzSvxS8xNVbJSSk4sTk5MSQ3LTC1X0lHyTS3OCEotVrIy0DOE84IS89KBSqMN9AwMdYxidZScE5MzUu2VrNISc4pTdZSCElMyS4F6jA1gbJgOIwMdEwOgBp%2F8tBKg%2BpKi0tRaAA%3D%3D)
